### PR TITLE
MultiError: Refactored MultiError for more concise and safe usage.

### PR DESCRIPTION
**Pull request by bwplotka**
* Fewer lines of code to use combined error safely.
* Goland IDE was marking every usage of the old MultiError "potential nil" error
* It was easy to forget using Err() when the error was returned, now it's safely assured on compile time.

NOTE: Potentially I would rename the package to `merrors`. (: In different PR.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: cmd/promtool/tsdb.go [hunk 1]
```diff
@@ -40,8 +40,6 @@ import (
 	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
 )
 
-var merr tsdb_errors.MultiError
```
Comment (bwplotka) [inline cmd/promtool/tsdb.go:43]: wth was that

File: tsdb/errors/errors.go [hunk 2]
```diff
@@ -17,21 +17,60 @@ package errors
...
+	errs multiError
+}
+
+// Error returns a concatenated string of the contained errors.
+func (es nonNilMultiError) Error() string {
 	var buf bytes.Buffer
 
-	if len(es) > 1 {
-		fmt.Fprintf(&buf, "%d errors: ", len(es))
+	if len(es.errs) > 1 {
```
Comment (ArthurSens) [inline tsdb/errors/errors.go:69]: if `nonMultiError` is only created when having at least 1 error, is this validation really necessary?

File: tsdb/errors/errors.go [hunk 3]
```diff
@@ -41,22 +80,11 @@ func (es MultiError) Error() string {
...
-	} else {
-		*es = append(*es, err)
-	}
-}
-
-// Err returns the error list as an error or nil if it is empty.
-func (es MultiError) Err() error {
-	if len(es) == 0 {
-		return nil
+// CloseAll closed all given closes with all errors recorded. All closers are closed.
```
Comment (ArthurSens) [inline tsdb/errors/errors.go:83]: Probably because I'm new here, but this comment is quite confusing :open_mouth:

**Review by ArthurSens — COMMENTED** [review]
Just one small thing I noticed

File: tsdb/errors/errors.go (hunk 2 above)
Comment (bwplotka) [reply tsdb/errors/errors.go:69]: Good point (: I guess not :+1:

File: tsdb/errors/errors.go (hunk 3 above)
Comment (bwplotka) [reply tsdb/errors/errors.go:83]: no it's a typo will fix (:

File: tsdb/errors/errors.go (hunk 2 above)
Comment (ArthurSens) [reply tsdb/errors/errors.go:69]: oops, I read if len > 0
probably need glasses lol

File: tsdb/querier.go [hunk 4]
```diff
@@ -96,12 +96,12 @@ func (q *blockBaseQuerier) Close() error {
 	if q.closed {
 		return errors.New("block querier already closed")
 	}
-	var merr tsdb_errors.MultiError
-	merr.Add(q.index.Close())
-	merr.Add(q.chunks.Close())
-	merr.Add(q.tombstones.Close())
-	q.closed = true
-	return merr.Err()
+	defer func() { q.closed = true }()
```
Comment (brian-brazil) [inline tsdb/querier.go:99]: Why did you change this logic?

File: tsdb/errors/errors.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 The etcd Authors
+// Copyright 2020 The Prometheus Authors
```
Comment (brian-brazil) [inline tsdb/errors/errors.go:1]: Why are you changing the author?

File: tsdb/errors/errors.go [hunk 6]
```diff
@@ -17,21 +17,60 @@ package errors
 import (
 	"bytes"
 	"fmt"
+	"io"
 )
 
-// The MultiError type implements the error interface, and contains the
-// Errors used to construct it.
-type MultiError []error
+// multiError type allows to combine multiple errors into one.
```
Comment (brian-brazil) [inline tsdb/errors/errors.go:23]: allows combining

File: tsdb/errors/errors.go [hunk 7]
```diff
@@ -17,21 +17,60 @@ package errors
...
-// Returns a concatenated string of the contained errors
-func (es MultiError) Error() string {
+// NewMulti returns multiError with provided errors added if not nil.
+func NewMulti(errs ...error) multiError { // nolint:golint
+	m := multiError{}
+	m.Add(errs...)
+	return m
+}
+
+// Add adds the error to the error list if it is not nil.
```
Comment (brian-brazil) [inline tsdb/errors/errors.go:33]: errs is a list though, not a single error

File: tsdb/errors/errors.go (hunk 5 above)
Comment (bwplotka) [reply tsdb/errors/errors.go:1]: Because we rewrote completely the code? (:

File: tsdb/querier.go (hunk 4 above)
Comment (bwplotka) [reply tsdb/querier.go:99]: Logic is exactly the same I believe :thinking:

File: tsdb/errors/errors.go (hunk 5 above)
Comment (bwplotka) [reply tsdb/errors/errors.go:1]: happy to revert

File: tsdb/errors/errors.go (hunk 6 above)
Comment (brian-brazil) [reply tsdb/errors/errors.go:23]: This comment wasn't addressed.

File: tsdb/errors/errors.go [hunk 8]
```diff
@@ -17,21 +17,60 @@ package errors
...
-func (es MultiError) Error() string {
+// NewMulti returns multiError with provided errors added if not nil.
+func NewMulti(errs ...error) multiError { // nolint:golint
+	m := multiError{}
+	m.Add(errs...)
+	return m
+}
+
+// Add adds single or many errors to the error list. Each error is added only if not nil.
+// If the error is a nonNilMultiError error, errors from passed multi errors are combine in the
```
Comment (brian-brazil) [inline tsdb/errors/errors.go:34]: This doesn't parse

File: tsdb/querier.go (hunk 4 above)
Comment (brian-brazil) [reply tsdb/querier.go:99]: It feels a bit weird to use a defer like this, it'd be better to make it explicit I think.

Comment (bwplotka) [reply tsdb/querier.go:99]: :+1:

File: tsdb/errors/errors.go (hunk 8 above)
Comment (bwplotka) [reply tsdb/errors/errors.go]: What about now?

**Comment by bwplotka** [comment]
Thanks for quick review round, rdy for next one :hugs:

**Review by brian-brazil — COMMENTED** [review]
:+1:

**Review by kakkoyun — APPROVED** [review]
Nice 👍

**Review by codesome — APPROVED** [review]
<3

**Comment by codesome** [comment]
@bwplotka can you fix the conflict?
